libamqpprox/amqpprox_connectionlimitermanager.cpp

```diff
@@ -0,0 +1,266 @@
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::setAlarmOnlyDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultAlarmOnlyConnectionRateLimit = defaultConnectionRateLimit;
+    // To update new default connection rate limit for all the vhosts,
+    // by removing old already set default connection rate limiters
+    for (auto &limiter : d_alarmOnlyConnectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultAlarmOnlyConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::removeConnectionRateLimiter(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_connectionRateLimitersPerVhost.erase(vhostName);
+    if (d_defaultConnectionRateLimit) {
+        d_connectionRateLimitersPerVhost[vhostName] = {
+            false,
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                d_defaultConnectionRateLimit)};
+    }
+}
+
+void ConnectionLimiterManager::removeAlarmOnlyConnectionRateLimiter(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_alarmOnlyConnectionRateLimitersPerVhost.erase(vhostName);
+    if (d_defaultAlarmOnlyConnectionRateLimit) {
+        d_alarmOnlyConnectionRateLimitersPerVhost[vhostName] = {
+            false,
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                d_defaultAlarmOnlyConnectionRateLimit)};
+    }
+}
+
+void ConnectionLimiterManager::removeDefaultConnectionRateLimit()
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultConnectionRateLimit = 0;
+    for (auto it = d_connectionRateLimitersPerVhost.cbegin();
+         it != d_connectionRateLimitersPerVhost.cend();) {
+        if (!it->second.first) {
+            d_connectionRateLimitersPerVhost.erase(it++);
+        }
+        else {
+            ++it;
+        }
+    }
+}
+
+void ConnectionLimiterManager::removeAlarmOnlyDefaultConnectionRateLimit()
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultAlarmOnlyConnectionRateLimit = 0;
+    for (auto it = d_alarmOnlyConnectionRateLimitersPerVhost.cbegin();
+         it != d_alarmOnlyConnectionRateLimitersPerVhost.cend();) {
+        if (!it->second.first) {
+            d_alarmOnlyConnectionRateLimitersPerVhost.erase(it++);
+        }
+        else {
+            ++it;
+        }
+    }
+}
+
+bool ConnectionLimiterManager::allowNewConnectionForVhost(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    if (d_alarmOnlyConnectionRateLimitersPerVhost.find(vhostName) ==
```

Comment:
would be nice to have a template function here that does this for each class since the code is duplicated

e.g. `maybePopulateDefaultLimiters(vhost, container<T> limiterContainer, T default limiter)`

then this simplifies to 
`maybePopulateDefaultLimiters(vhost, d_alarmOnlyConnectionRateLimitersPerVhost, d_defaultAlarmOnlyConnectionRateLimit);`
`maybePopulateDefaultLimiters(vhost, d_connectionRateLimitersPerVhost, d_defaultConnectionRateLimit);`